scipy/interpolate/_rgi_cython.pyx

```diff
@@ -0,0 +1,181 @@
+import numpy as np
+cimport numpy as np
+cimport cython
+
+np.import_array()
+
+import itertools
+
+
+cdef extern from "numpy/npy_math.h":
+    double nan "NPY_NAN"
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+cdef int find_interval_ascending(const double *x,
+                                 size_t nx,
+                                 double xval,
+                                 int prev_interval=0,
+                                 bint extrapolate=1) nogil:
+    """
+    Find an interval such that x[interval] <= xval < x[interval+1]. Assuming
+    that x is sorted in the ascending order.
+    If xval < x[0], then interval = 0, if xval > x[-1] then interval = n - 2.
+
+    Parameters
+    ----------
+    x : array of double, shape (m,)
+        Piecewise polynomial breakpoints sorted in ascending order.
+    xval : double
+        Point to find.
+    prev_interval : int, optional
+        Interval where a previous point was found.
+    extrapolate : bint, optional
+        Whether to return the last of the first interval if the
+        point is out-of-bounds.
+
+    Returns
+    -------
+    interval : int
+        Suitable interval or -1 if nan.
+
+    """
+    cdef int interval, high, low, mid
+    cdef double a, b
+
+    a = x[0]
+    b = x[nx-1]
+
+    interval = prev_interval
+    if interval < 0 or interval >= nx:
+        interval = 0
+
+    if not (a <= xval <= b):
```

Comment:
If you have a prior assumption about the most probable branch, it is better to put it as the first.